### PR TITLE
Replace 'cstderr.rawWrite' with 'writeToStdErr'

### PR DESCRIPTION
**Pull request by ghost**
Use a procedure that is easy to replace on platforms without native UNIX file descriptors.

(#17268 reattempt)

## future work
- [ ] make writeToStdErr work in VM via a magic/vmops (without resorting to --experimental:compiletimeFFI)
- [ ] ditto for js backend, where we can map it to console.log

### CODE REVIEW

**Comment by ghost** [comment]
I'm not sure if these failed checks are related?

File: lib/system/mmdisp.nim [hunk 1]
```diff
@@ -45,7 +45,7 @@ else:
 
 proc raiseOutOfMem() {.noinline.} =
   if outOfMemHook != nil: outOfMemHook()
-  cstderr.rawWrite("out of memory\n")
+  writeToStdErr("out of memory")
```
Comment (timotheecour) [inline lib/system/mmdisp.nim]:
```suggestion
  writeToStdErr("out of memory\n")
```

(why was this changed?)

File: lib/system/fatal.nim [hunk 2]
```diff
@@ -42,7 +48,7 @@ elif (defined(nimQuirky) or defined(nimPanics)) and not defined(nimscript):
       add(buf, " [")
       add(buf, name exceptn)
       add(buf, "]\n")
-      cstderr.rawWrite buf
+      writeFatalErr(buf.cstring, buf.len)
```
Comment (timotheecour) [inline lib/system/fatal.nim:51]:
```suggestion
      writeToStdErr buf
```
I don't think there's a need to introduce `writeFatalErr`, instead `writeToStdErr` should be made to work cross platform

Comment (ghost) [reply lib/system/fatal.nim:51]: I would prefer to reuse `writeToStdErr` but I couldn't find an elegant way to do it, `writeToStdErr` is mostly used in files that are `include`ed into `system.nim`, and as I recall `system/fatal` is imported or included in multiple locations.

File: lib/system.nim [hunk 3]
```diff
@@ -551,6 +551,10 @@ when notJSnotNims and not defined(nimSeqsV2):
 when notJSnotNims:
   include "system/hti"
 
+  proc writeToStdErr(msg: cstring, length: int)
+  proc writeToStdErr(msg: cstring)
```
Comment (timotheecour) [inline lib/system.nim:555]: how about:
```nim
  proc writeToStdErr(msg: cstring) {.inlnie.} =
    writeToStdErr(msg, msg.len)
```

that way, no need to define this overload `proc writeToStdErr(msg: cstring)` on each platform

File: lib/system/mmdisp.nim (hunk 1 above)
Comment (ghost) [reply lib/system/mmdisp.nim]: Must have just been for style, the next line is `quit` so the buffer should flush regardless.

File: lib/system.nim [hunk 4]
```diff
@@ -551,6 +551,10 @@ when notJSnotNims and not defined(nimSeqsV2):
...
+  proc writeToStdErr(msg: string)
```
Comment (timotheecour) [inline lib/system.nim:556]: ditto, this could be defined directly as:
```nim
  proc writeToStdErr(msg: string) {.inline.} =
    writeToStdErr(msg.cstring, msg.len)
```

that way, the only proc to define on each platform is `proc writeToStdErr(msg: cstring, length: int)`

File: lib/system/mmdisp.nim (hunk 1 above)
Comment (timotheecour) [reply lib/system/mmdisp.nim]: adding a newline isn't the same as flushing; this changes semantics and resulting stderr output

File: lib/system/fatal.nim (hunk 2 above)
Comment (timotheecour) [reply lib/system/fatal.nim:51]: i think it's fixable, once you fix the other points i can look into how to fix that

File: lib/system.nim [hunk 5]
```diff
@@ -551,6 +551,10 @@ when notJSnotNims and not defined(nimSeqsV2):
 when notJSnotNims:
   include "system/hti"
 
+  proc writeToStdErr(msg: cstring, length: int)
```
Comment (timotheecour) [inline lib/system.nim:554]: > I'm not sure if these failed checks are related?

most likely related, and you should be able to reproduce the bug locally with smthg like
`testament pcat dll` or similar (these are the failing tests in CI i see)

**Comment by ghost** [comment]
I'm not happy with this patch. Its hard to know where define `writeToStdErr` in system because its used in `include` files but depends on `ansi_c`. Would it be better to replace this with an internal compiler builtin instead? I'm willing to make the changes but with the sytem and hardware I'm using its difficult to run the tests.

**Comment by Araq** [comment]
Why not add it to `ansi_c.nim` then. Please not another built-in.

**Comment by ghost** [comment]
Would this only be valid for C-comptible targets? This is something mostly used in the garbage collector. If it was to go into `ansi_c` then I would want to change the name to something like `writeSystemError`, because it wouldn't be UNIX standard error for some platforms.

**Comment by Araq** [comment]
> Would this only be valid for C-comptible targets?

ansi_c.nim would contain the default implementation, you can re-define it with this pattern:

```nim

when not declared(writeError):
  proc writeError() = ...

```

**Comment by timotheecour** [comment]
ping @ehmry this is a good PR, please continue if you can!

**Comment by ghost** [comment]
I don't know the code here well enough to make this change in a way that improves its quality, and I don't think I'm in a position to catch regressions. It would be a welcome improvement, but not something that I have problem with at the moment.
